gossip/src/duplicate_shred.rs

```diff
@@ -84,35 +85,71 @@ pub enum Error {
+///     - Must be for the same slot
+///     - Must have the same `shred_type`
+///     - Must both sigverify for the correct leader
+///     - If `shred1` and `shred2` share the same index they must be not equal
+///     - If `shred1` and `shred2` do not share the same index and are data shreds
+///       verify that they indicate an index conflict. One of them must be the
+///       LAST_SHRED_IN_SLOT, however the other shred must have a higher index.
+///     - If `shred1` and `shred2` do not share the same index and are coding shreds
+///       verify that they have conflicting erasure metas
 fn check_shreds<F>(leader_schedule: Option<F>, shred1: &Shred, shred2: &Shred) -> Result<(), Error>
 where
     F: FnOnce(Slot) -> Option<Pubkey>,
 {
     if shred1.slot() != shred2.slot() {
-        Err(Error::SlotMismatch)
-    } else if shred1.index() != shred2.index() {
-        // TODO: Should also allow two coding shreds with different indices but
-        // same fec-set-index and mismatching erasure-config.
-        Err(Error::ShredIndexMismatch)
-    } else if shred1.shred_type() != shred2.shred_type() {
-        Err(Error::ShredTypeMismatch)
-    } else if shred1.payload() == shred2.payload() {
-        Err(Error::InvalidDuplicateShreds)
-    } else {
-        if let Some(leader_schedule) = leader_schedule {
-            let slot_leader =
-                leader_schedule(shred1.slot()).ok_or(Error::UnknownSlotLeader(shred1.slot()))?;
-            if !shred1.verify(&slot_leader) || !shred2.verify(&slot_leader) {
-                return Err(Error::InvalidSignature);
-            }
+        return Err(Error::SlotMismatch);
+    }
+
+    if shred1.shred_type() != shred2.shred_type() {
+        return Err(Error::ShredTypeMismatch);
+    }
+
+    if shred1.index() == shred2.index() && shred1.payload() == shred2.payload() {
+        return Err(Error::InvalidDuplicateShreds);
+    }
+
+    if shred1.index() != shred2.index() && shred1.shred_type() == ShredType::Data {
+        match (
+            shred1.index(),
+            shred1.last_in_slot(),
+            shred2.index(),
+            shred2.last_in_slot(),
+        ) {
+            (ix1, true, ix2, false) if ix1 > ix2 => return Err(Error::InvalidLastIndexConflict),
+            (ix1, false, ix2, true) if ix1 < ix2 => return Err(Error::InvalidLastIndexConflict),
+            (_, false, _, false) => return Err(Error::InvalidLastIndexConflict),
+            _ => (),
+        }
+    }
+
+    if shred1.index() != shred2.index() && shred1.shred_type() == ShredType::Code {
+        if shred1.fec_set_index() != shred2.fec_set_index() {
+            return Err(Error::InvalidErasureMetaConflict);
+        }
+        let erasure_meta =
+            ErasureMeta::from_coding_shred(shred1).expect("Shred1 should be a coding shred");
+        if erasure_meta.check_coding_shred(shred2) {
+            return Err(Error::InvalidErasureMetaConflict);
```

Comment:
This is not technically exhaustive.
Any 2 shreds with different but overlapping erasure sets can be considered duplicate.
In particular doesn't even need to have the same `fec_set_index`.

Technically the overlapping erasure sets might result in non-duplicate shreds, but a leader shouldn't generate overlapping erasure sets unless malicious or buggy.

We don't currently identify those cases because we just lookup by the erasure set:
https://github.com/solana-labs/solana/blob/5de27d9a0/ledger/src/blockstore.rs#L1197-L1206
but I think that has to be fixed at some point.

Reply:
Good to know, right now i've only replicated the cases that we catch in blockstore. 
I created https://github.com/solana-labs/solana/issues/33037, and will extend the checks to catch these types of conflicts in a follow up pr.